lib/synapse/service_watcher/asg.rb

```diff
@@ -0,0 +1,139 @@
+require 'synapse/service_watcher/base'
+require 'aws-sdk'
+
+class Synapse::ServiceWatcher
+  class AsgWatcher < BaseWatcher
+
+    attr_reader :check_interval
+
+    def start
+      region = @discovery['aws_region'] || ENV['AWS_REGION']
+      log.info "Connecting to EC2 region: #{region}"
+
+      @asg_client = AWS::AutoScaling::Client.new(
+          region:            region,
+          access_key_id:     @discovery['aws_access_key_id']     || ENV['AWS_ACCESS_KEY_ID'],
+          secret_access_key: @discovery['aws_secret_access_key'] || ENV['AWS_SECRET_ACCESS_KEY'] )
+
+      @ec2_client = AWS::EC2::Client.new(
+          region:            region,
+          access_key_id:     @discovery['aws_access_key_id']     || ENV['AWS_ACCESS_KEY_ID'],
+          secret_access_key: @discovery['aws_secret_access_key'] || ENV['AWS_SECRET_ACCESS_KEY'] )
+
+      @check_interval = @discovery['check_interval'] || 15.0
+
+      log.info "synapse: asg watcher looking for instances in asg #{@discovery["asg_name"]}"
+
+      @watcher = Thread.new { watch }
```

Comment:
Just a heads up there is a flaw with how the ec2tag watcher and other "watch in a Thread" watchers are implemented. Because this returns immediately from the thread creation synapse might proceed to restart HAProxy before watch has actually been called. The zookeeper watcher gets around this because it uses a callback system and waits for the callback to finish once before returning.

I'm not saying you have to change it, but I just want you to be aware of the shortcoming. I'm super open to ideas for how to fix it.